<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Document</title>
</head>
<body>
  <script>
    // const promise = new Promise((resolve, reject) => {
    //   console.log(1)
    //   resolve()
    //   console.log(2)
    // })
    // promise.then(() => {
    //   console.log(3)
    // })
    // console.log(4)
    // 1 2 4 3

    // const promise1 = new Promise((resolve, reject) => {
    //   setTimeout(() => {
    //     resolve('success')
    //   }, 1000)
    // })
    // const promise2 = promise1.then(() => {
    //   throw new Error('error!!!')
    // })

    // console.log('promise1', promise1)
    // console.log('promise2', promise2)

    // setTimeout(() => {
    //   console.log('promise1', promise1)
    //   console.log('promise2', promise2)
    // }, 2000)
    // // pending pending fulfilled rejected

    // const promise = new Promise((resolve, reject) => {
    //   resolve('success1')
    //   reject('error')
    //   resolve('success2')
    // })

    // promise
    //   .then((res) => {
    //     console.log('then: ', res)
    //   })
    //   .catch((err) => {
    //     console.log('catch: ', err)
    //   })
    // then: success1

    // Promise.resolve(1)
    //   .then((res) => {
    //     console.log(res)
    //     return 2
    //   })
    //   .catch((err) => {
    //     return 3
    //   })
    //   .then((res) => {
    //     console.log(res)
    //   })
    // 1 2
    // const promise = new Promise((resolve, reject) => {
    //   setTimeout(() => {
    //     console.log('once')
    //     resolve('success')
    //   }, 1000)
    // })

    // const start = Date.now()
    // promise.then((res) => {
    //   console.log(res, Date.now() - start)
    // })
    // promise.then((res) => {
    //   console.log(res, Date.now() - start)
    // })
    // once
    // success 1002
    // success 1002

    // Promise.resolve()
    //   .then(() => {
    //     return new Error('error!!!')
    //   })
    //   .then((res) => {
    //     console.log('then: ', res)
    //   })
    //   .catch((err) => {
    //     console.log('catch: ', err)
    //   })
    // catch: error!!!

    // const promise = Promise.resolve()
    //   .then(() => {
    //     return promise
    //   })
    // promise.catch(console.error)

    Promise.resolve()
      .then(function success (res) {
        throw new Error('error')
      }, function fail1 (e) {
        console.error('fail1: ', e)
      })
      .catch(function fail2 (e) {
        console.error('fail2: ', e)
      })
      // fail2: error
  </script>
</body>
</html>